NATS interface for WMAgent monitoring

vkuznet opened the pull request:
Fixes : there is no issues

#### Status
In development

#### Description
This PR provides a NATS interface for WMAgent/WMCore/ReqMgr2. It was defined as analogous to DashboardInterface and consists of two classes:
- NATSInterface which extract proper information from task/job
- NATSMonitor a generic monitor interface for WMAgent/WMCore
It also adds [NATSManager](https://github.com/dmwm/CMSMonitoring/blob/master/src/python/CMSMonitoring/NATS.py) to [Request.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/ReqMgr/Service/Request.py) and NATSMonitor to [StdBase](https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py). The former will allow to yield messages on ReqMgr workflow transitions, while later supposes to cover WMAgent production activity at a site.

To validate the functionality we need to run an external NATS subscriber which will listen to NATS server. The NATS subscriber tool can be obtained on AFS: /afs/cern.ch/user/v/valya/public/nats-sub and it can be run as following:
```
# to monitor reqmgr2 messages
./nats-sub -rootCAs=/path/certificates/CERN_CA.crt,/path/certificates/CERN_CA1.crt -t "cms.reqmgr2.>"

# to monitor WMA messages
./nats-sub -rootCAs=/path/certificates/CERN_CA.crt,/path/certificates/CERN_CA1.crt -t "cms.wma.>"
```

The tool requires rootCAs certificates which can be obtained from `ca.cern.ch` and placed elsewhere, as well as it requires `$HOME/.nats/cms-auth` file which contains login:password to authenticate with NATS server.

In case of ReqMgr2 monitoring we should simply add a new workflow to ReqMgr2 and monitor its transition changes, e.g. here are expected messages:
```
# the message is composed by TimeStamp Task OutputDatasets ProcessingString Priority Status
2019/12/10 16:07:30 valya_TaskChain_MC_Agent105_CDB16_pre_191210_160725_4377 ["/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-81X_mcRun2_asymptotic_v12-v20/GEN-SIM", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-81X_mcRun2_asymptotic_v12-v20/AODSIM", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-81X_mcRun2_asymptotic_v12-v20/DQMIO", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-81X_mcRun2_asymptotic_v12-v20/MINIAODSIM"] 81X_mcRun2_asymptotic_v12 600000 new
2019/12/10 16:07:30 valya_TaskChain_MC_Agent105_CDB16_pre_191210_160725_4377 []  0 assignment-approved
2019/12/10 16:07:30 valya_TaskChain_MC_Agent105_CDB16_pre_191210_160725_4377 ["/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-ProdQCD_Pt_3000_3500_13_TaskChain_MC_pre_VK_TEST-v11/GEN-SIM", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-RECOPRODUP15_TaskChain_MC_pre_VK_TEST-v11/AODSIM", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-RECOPRODUP15_TaskChain_MC_pre_VK_TEST-v11/DQMIO", "/RelValProdQCD_Pt_3000_3500_13/CMSSW_8_1_0-RECOPRODUP15_TaskChain_MC_pre_VK_TEST-v11/MINIAODSIM"] ["DIGIUP15PROD1", "ProdQCD_Pt_3000_3500_13", "RECOPRODUP15"] 0 assigned
```

In case of WMA monitoring we need to assign a job to WMAgent and monitor its progress using `cms.wma.>` end-point of NATS subscriber (see example of `nats-sub` call above). In that case the messages on that end-point should contains site, job, exitcode, step information. These attributes are defined in NATSMonitor (see files in this PR).

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
This PR fixes the following issues: #9449, #9448 

#### External dependencies / deployment changes
[CMSMonitoring](github.com/dmwm/CMSMonitoring) package provides all required dependencies.

vkuznet committed:
NATS interface for WMAgent monitoring

vkuznet commented:
Alan, I created this PR as placeholder for NATS support.

I followed DashboardInterface and persistent task/job info. So far I extracted task, job, campaign attributes but we will need to complement NATS messages with Site and proper exit codes. Please advise how to extract those from task/job/workload info.

Finally, I don't know where you run DashboardMonitor but I would presume this will be run at the same place. My understanding that NATSMonitor will be accompany each task and its methods (steps, jobs) will be called at proper time by the agents.

vkuznet commented:
The complain of pylint: E0611, line 12: No name 'NATS' in module 'CMSMonitoring' possibly comes from the fact that WMCore jenkins does not have CMSMonitoring package in their PYTHONPATH. Apart from this error I don't see anything else which require my attention to pylint.

amaltaro commented:
Actually the problem is that the CMSMonitoring package we rely on still doesn't have the NATS module.
So we need to update the cmsmonitoring spec to the latest stable release (one which includes NATS), then update and build a new wmagent-dev release.

vkuznet commented:
Ahh, I didn't know that for jenkins you need that.

I submitted the changes:
- CMSMonitoring, verison 0.2.7
- new cmsdist spec, https://github.com/cms-sw/cmsdist/pull/5362
- pypi: changes should appear soon (they have some latency with new tags)

There is no rush here anyway. I posted it for completeness.

On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> Actually the problem is that the CMSMonitoring package we rely on still doesn't have the NATS module.
> So we need to update the cmsmonitoring spec to the latest stable release (one which includes NATS), then update and build a new wmagent-dev release.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/9424#issuecomment-555209668

vkuznet commented:
Few updates:
- cms-nats server is now fully encrypted and firewall is opened
- the CMSMonitoring version which provides TLS support, etc., is 0.3.1, even though 0.3.0 also work. The corresponding PR: https://github.com/cms-sw/cmsdist/pull/5379

For completeness here I'm posting basic code how to use NATSManager:
```
#!/usr/bin/env python

# system modules
import os
import ssl
import sys
import argparse

# CMSMonitoring modules
from CMSMonitoring.NATS import NATSManager


class OptionParser():
    def __init__(self):
        "User based option parser"
        self.parser = argparse.ArgumentParser(prog='PROG')
        self.parser.add_argument("--server", action="store",
                                 dest="server", default="cms-nats.cern.ch",
                                 help="NATS server, default cms-nats.cern.ch")
        self.parser.add_argument("--auth", action="store",
                                 dest="auth", default="",
                                 help="NATS auth credentials")
        self.parser.add_argument("--subject", action="store",
                                 dest="subject", default="",
                                 help="subject NATS end-point")
        self.parser.add_argument("--verbose", action="store_true",
                                 dest="verbose", default=False,
                                 help="verbose output")

def main():
    "Main function"
    optmgr  = OptionParser()
    opts = optmgr.parser.parse_args()

    # publish a message
    server = opts.server   # e.g. cms-nats.cern.ch
    auth = opts.auth       # e.g. /path/cms-auth
    subject = opts.subject # e.g. T2_US_Test

    # adjust server with authentication string
    if auth:
        if os.path.exists(auth):
            auth = '{}@'.format(open(auth).read().replace('\n', ''))
    else:
        fname = '{}/.nats/cms-auth'.format(os.getenv('HOME'))
        if os.path.exists(fname):
            auth = '{}@'.format(open(fname).read().replace('\n', ''))
    server = "tls://{}{}:4222".format(auth, server)

    # get NATSManager instance
    mgr = NATSManager(server, topics=[subject], default_topic='cms-test.Test')
    # generate some JSON docs, e.g.
    docs = []
    site = 'T2_US_Test'
    for idx in range(3):
        dataset = '/a/b/tier-{}'.format(idx)
        task = 'task-{}'.format(idx)
        campaign = 'campaign-{}'.format(idx)
        doc = {'site':site, 'dataset':dataset, 'task':task, 'campaign':campaign}
        docs.append(doc)
    mgr.publish(docs)

if __name__ == '__main__':
    main()
```

vkuznet commented:
I want to mentioned here that I successfully tested NATS with DBS, see https://github.com/dmwm/DBS/pull/617

Here, I need to know how to extract the following parts:
- exit code
- log file
- site
- job locality
and I need to know  where to place NATS interface that it will properly yield job status.

vkuznet commented:
@amaltaro , could you please update wmagent-dev release to include new version of cmsmonitoring spec since pylint failed to find NATS from CMSMonitoring module.

I think, I implemented basic logic for all use-cases: #9448, #9449, #9452 in this PR, and I can proceed with integration tests. Feel free to make any comment.

amaltaro commented:
Valentin, I have just triggered the build of new docker images. We can check these unit tests again tomorrow.
This PR will require a closer look, so it might take a while until I get back to you here. Thanks for putting all these in place though.

vkuznet commented:
Just tested the code in private VM, it works!!!!

I submitted simple workflow and ReqMgr2 was happily published messages to NATS server, see
```
./nats-sub -rootCAs=/afs/cern.ch/user/v/valya/private/certificates/CERN_CA.crt,/afs/cern.ch/user/v/valya/private/certificates/CERN_CA1.crt -t "cms.reqmgr2.>"
Listening on [cms-nats.cern.ch/cms.reqmgr2.>]
2019/12/10 00:26:38 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 600000 new
2019/12/10 00:26:38 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 600000 new
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 600000 new
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assignment-approved
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assignment-approved
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assignment-approved
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assigned
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assigned
2019/12/10 00:26:39 valya_TaskChain_MC_Agent105_CDB16_forceComplete_191210_002636_7677 0 assigned
```
I still need to polish the code and remove duplicate publications (find where is a proper place to publish the stuff), but integration wise it works like a charm since I can see the messages in NATS subscriber.

vkuznet commented:
@amaltaro , Alan, I polished this PR and it is ready for review. Here is a list of tasks I accomplished:
- I deployed this PR to private VM and tests that ReqMgr2 sends messages to NATS and I can see them
- I changed code to satisfy pylint/pep requirements, but I didn't touched reported issues which are not related to this PR
- I didn't test NATSMonitor since I don't know how, you should either provide instructions and/or run your tests. The NATSMonitor changes are similar to those I made in ReqMgr2 therefore I don't expect any integration issues. We only need to test when monitors kicks-in in WMAgent workflows
- For ReqMgr2 I addressed use-case https://github.com/dmwm/WMCore/issues/9449
- The NATSMonitor addresses use-case https://github.com/dmwm/WMCore/issues/9448
- the use case https://github.com/dmwm/WMCore/issues/9452 is not addressed yet since I don't know where to locate proper information and I don't know if it should be part of NATSMonitor or not.

Therefore I suggest that you review the PR and suggest how we can move forward with it. My suggestion to roll-over/test gradually these changes.

amaltaro commented:
Thanks for the update, Valentin.
Before I review it, can you please:
* update the your first comment in this PR with: 
  * which issues it fixes
  * provide a description of what exactly is this PR for (and what it provides)
  * and for anyone that will need to validate it, steps on how it can be tested (e.g., how you subscribe to a given workflow or whatever, and see updates and etc).

vkuznet commented:
Done, please see description. When you'll be ready to test you'll need to obtain from me login/password for nats subscriber. If you want I can show you live demo in slack/vidyo.

amaltaro commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -487,6 +499,32 @@ def _handleOnlyStateTransition(self, workload, request_args, dn):
         report = self.reqmgr_db_service.updateRequestStatus(workload.name(), req_status, dn)
         return report
 
+    def _updateNATS(self, workload, request_args):
```
Valentin, not really a review, but just a comment on something that caught my attention.

Why did you chose this module to implement these updates? I think StdBase, same as done for old dashboard:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L248

would make much more sense.

vkuznet replied:
We need two monitoring, one for ReqMgr transitions and one for workflows. This was requested by McM group. The former is implemented in ReqMgr, the later in WMRuntime.

I asked originally about where it should be done, since I didn't heard from you I implemented where I new it will be possible. You better know the entire system and I would rely on your expertise to guide where the changes should be added.

The change I made in ReqMgr I tested with reqmgr2 service and it seems to work fine for transitions. If you think it should be relocated please provide full details.

amaltaro commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -36,6 +40,14 @@ def __init__(self, app, api, config, mount):
+        # initialize NATS if requested
+        self.nats = None
+        if hasattr(config, 'use_nats') and config.use_nats and hasattr(config, 'nats_server'):
+            topic = 'cms.reqmgr2'
```
Should we make it configurable? Or it's very likely that this is not going to change in the next few years?

vkuznet replied:
This is the *main* default topic. I'd like to keep them fixed, `cms.reqmgr2` for ReqMgr2, `cms.dbs` for DBS, `cms.wmarchive` for WMArchive. Therefore I don't see needs to configure it here.

vkuznet replied:
within the main topic the messages may be divided into sub-topics, e.g. I already have `cms.wmarchive.site`, `cms.wmarchive.exitCodes`, etc. Therefore I would expect that in ReqMgr2 we'll define sub-topics too and those may be configurable.

vkuznet commented:
@alan, can you please review this PR now and possibly include it in this cmsweb release cycle. I understand that Agents are upgraded separately, but this PR also contains code changes for ReqMgr2 where we can enable NATS and ReqMgr we update in cmsweb release cycle.

amaltaro commented on src/python/WMCore/WMRuntime/Monitors/NATSMonitor.py:
```diff
@@ -0,0 +1,111 @@
+
+# system modules
+import logging
+import traceback
```
no need to import traceback. Just use `logging.exception` instead (there are a few cases below)

vkuznet replied:
ok, done

amaltaro commented on src/python/WMCore/WMRuntime/NATSInterface.py:
```diff
@@ -0,0 +1,129 @@
+from CMSMonitoring.NATS import NATSManager
+
+
+class NATSInfo(object):
```
Should we rename it to something like `NATSPublisher`? The file is named as `NATSInterface`, which clearly it's not! NATSInfo doesn't seem to be the best choice either.

vkuznet replied:
For this round, I changed NATSInfo to NATSPublisher, but I kept NATSInterface as a module name, since Publisher may be just one of the interfaces. If you still be in favor of new name we can rename it.

amaltaro commented on src/python/WMCore/WMRuntime/NATSInterface.py:
```diff
@@ -0,0 +1,129 @@
+from WMCore.WMRuntime.Bootstrap import getSyncCE
+
+# CMSMonitoring modules
+from CMSMonitoring.NATS import NATSManager
```
Importing from CMSMonitoring from the worker node(?) I'm actually surprised that it's in the job environment...

vkuznet replied:
NATS codebase belongs to CMSMonitoring repository, that's why is the import.

amaltaro commented on src/python/WMCore/WMRuntime/NATSInterface.py:
```diff
@@ -0,0 +1,129 @@
+"""
+NATSInfo provides wrapper around NATS to send WMAgent messages
+"""
+from __future__ import division
```
these two can be grouped

vkuznet replied:
done

amaltaro commented on src/python/WMCore/WMRuntime/NATSInterface.py:
```diff
@@ -0,0 +1,129 @@
+
+        # NATS manager
+        self.nats = NATSManager(server, topics=topics, attrs=attrs, default_topic='cms.wma')
+        print("WMAgent NATS: %s" % self.nats)
```
Use the logging module instead.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/WMSpec/StdSpecs/StdBase.py:
```diff
@@ -243,6 +243,10 @@ def addRuntimeMonitors(self, task):
         monitoring.PerformanceMonitor.softTimeout = softTimeout
         monitoring.PerformanceMonitor.hardTimeout = hardTimeout
+        monitoring.section_("NATSMonitor")
+        monitoring.NATSMonitor.maxPSS = maxpss
```
These attributes should actually be something needed for connecting to the server (url, port, etc), just as done above for the DashboardMonitor (not PerformanceMonitor).

amaltaro reviewed (changes requested):
Valentin, sorry for the belated review, at least it's now complete.
I made many comments to the code, please check them out and provide the changes and/or discuss any questions or things that I might have misunderstood.

vkuznet commented on src/python/WMCore/WMRuntime/Monitors/NATSMonitor.py:
```diff
@@ -0,0 +1,111 @@
+        attrs = args.get('attrs', None)
+        self.mgr = NATSInfo(task, job, server, topics, attrs)
+
+    def shutdown(self):
```
ok, done

vkuznet commented on src/python/WMCore/WMRuntime/Monitors/NATSMonitor.py:
```diff
@@ -0,0 +1,111 @@
+            args = {}
+        server = args.get('server', '')
+        if not server:
+            raise Exception("Please provide valid NATS server")
```
ok, done

vkuznet reviewed (commented):
Alan, I applied all changes you requested

vkuznet commented:
Alan, I managed to get pylint, pep8, python3 tests. For pylint I only fixed issues related to this PR, but I didn't fixed previous issues with code itself, e.g. Request.py has pylint errors which is not related to this PR. Please review this PR again.

amaltaro commented:
@todor-ivanov as we just chatted over slack. Last time I talked to Valentin, the future usability of this PR wasn't very clear to me. Given that it's also not of paramount importance, I'd suggest to wait until March (until I'm back) and then we can resume this discussion.

sharad1126 commented:
@amaltaro @todor-ivanov I think we need to push this PR to a little more urgent priority as this is something that would really help PnR, site support and also site admins to debug issues faster than usual.

vkuznet commented:
@amaltaro , @todor-ivanov , what is the status of this PR? This PR is already 1/2 year old and we still didn't merge it. Last time I spoke with data-ops (Sharad and Jen) they were very enthusiastic to see it implemented.

amaltaro commented:
Valentin, we have quickly discussed with P&R about this NATS based monitoring. They are going to prepare a document with use cases and requirements from their side. 

With that in hands, we can have a proper discussion and get to a list of requirements to be implemented within WMCore.

amaltaro commented:
Given that I have a couple of long standing action items on this topic, let me try to get them done.

@vkuznet This PR has now been inactive for more than a year, there seems to be no consensus and/or effort from other projects (McM/P&R) to adopt NATS as a monitoring tool (and Justinas says the current workflow monitoring mechanism is in a good shape) and adopting a new workflow monitoring technology isn't in our priorities for this year. 

With all that said, I'm reluctant to resume working on this, especially because it adds a new dependency to the WMCore code which is very cloudy at the moment, as well as its future adoption by those monitoring production workflows. Thus, I'm inclined in closing this PR as is, if we decide to adopt it in the future, we can always reopen it anyways and resume working on it.

FYI @klannon @KatyEllis

vkuznet commented:
@amaltaro , that's fine with me. Closing the PR then.
